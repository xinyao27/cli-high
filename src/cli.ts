import path from 'node:path'
import fs from 'node:fs/promises'
import process from 'node:process'
import { hideBin } from 'yargs/helpers'
import yargs from 'yargs'
import { intro, log, note } from '@clack/prompts'
import { dim, inverse, magentaStylize, redStylize } from 'xycolors'
import pkgJson from '../package.json'
import { highlight } from './index'

function header() {
  console.log('\n')
  intro(`✨ ${magentaStylize(`${pkgJson.name} `)}${dim(`v${pkgJson.version}`)}`)
}

const instance = yargs(hideBin(process.argv))
  .scriptName('cli-high')
  .usage('')
  .command(
    '*',
    'Outputs a file or STDIN input with syntax highlighting',
    (args) =>
      args
        .option('showLineNumbers', {
          alias: 's',
          description: 'Show line numbers',
          type: 'boolean',
        })
        .help(),
    async (args) => {
      const { _, showLineNumbers } = args
      header()

      try {
        const filePath = path.join(process.cwd(), _[0].toString())
        const code = await fs.readFile(filePath, 'utf-8')
        const highlighted = highlight(code, { showLineNumbers })
        note(highlighted)
      } catch (error) {
        log.error(inverse(redStylize(' Failed to read file')))
        log.error(redStylize(`✘ ${String(error)}`))
        process.exit(1)
      }
    },
  )
  .showHelpOnFail(false)
  .alias('h', 'help')
  .version('version', pkgJson.version)
  .alias('v', 'version')

// eslint-disable-next-line no-unused-expressions
instance.help().argv
