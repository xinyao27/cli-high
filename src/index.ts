import { highlight as highlightAsHtml } from 'sugar-high'
import { black, cyan, dim, green, magenta, red } from 'yoctocolors'

export interface HighlightOptions {
  showLineNumbers?: boolean
}
export function highlight(code: string, options: HighlightOptions = { showLineNumbers: false }) {
  const html = highlightAsHtml(code)
  const regex = /<span class="sh__line">(.*)<\/span>/g

  let match
  let result = ''
  let i = 1

  // eslint-disable-next-line no-cond-assign
  while ((match = regex.exec(html)) !== null) {
    const innerMatches = [...match[1].matchAll(/<span class="([^"]+)" style="color: ([^"]+)">(.*?)<\/span>/gs)]
    result += options.showLineNumbers ? `${dim(`${i}`.padEnd(2, ' '))} ` : ''
    innerMatches.forEach((innerMatch) => {
      const classType = innerMatch[1]
      const text = decode(innerMatch[3])

      switch (classType) {
        case 'sh__token--identifier':
          result += black(text)
          break
        case 'sh__token--keyword':
          result += red(text)
          break
        case 'sh__token--string':
          result += green(text)
          break
        case 'sh__token--class':
          result += cyan(text)
          break
        case 'sh__token--property':
          result += cyan(text)
          break
        case 'sh__token--entity':
          result += magenta(text)
          break
        case 'sh__token--jsxliterals':
          result += green(text)
          break
        case 'sh__token--sign':
          result += dim(text)
          break
        case 'sh__token--comment':
          result += dim(text)
          break
        default:
          result += text
          break
      }
    })
    result += '\n'
    i++
  }

  return result
}

function decode(str: string) {
  return str
    .replace('&amp;', '&')
    .replace('&lt;', '<')
    .replace('&gt;', '>')
    .replace('&quot;', '"')
    .replace('&#039;', "'")
}
