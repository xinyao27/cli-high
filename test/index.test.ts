/* eslint-disable no-console */
import path from 'node:path'
import { describe, expect, it } from 'bun:test'
import { highlight } from '../src/index'

const classjs = await Bun.file(path.resolve(__dirname, '../example/class.js')).text()
const regexjs = await Bun.file(path.resolve(__dirname, '../example/regex.js')).text()
const jsx = await Bun.file(path.resolve(__dirname, '../example/jsx.js')).text()

describe('should work', () => {
  it('class', () => {
    const result = highlight(classjs)
    console.log(result)
    expect(result).toMatchSnapshot()
  })

  it('regex', () => {
    const result = highlight(regexjs)
    console.log(result)
    expect(result).toMatchSnapshot()
  })

  it('jsx', () => {
    const result = highlight(jsx)
    console.log(result)
    expect(result).toMatchSnapshot()
  })

  it('jsx with line numbers', () => {
    const result = highlight(jsx, { showLineNumbers: true })
    console.log(result)
    expect(result).toMatchSnapshot()
  })
})
