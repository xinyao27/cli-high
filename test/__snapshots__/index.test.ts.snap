// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should work klassjs 1`] = `
"[2m/**[22m
[2m* @param {string} names[22m
[2m* @return {Promise<string[]>}[22m
[2m*/[22m
[31masync[39m [31mfunction[39m [30mnotify[39m[2m([22m[30mnames[39m[2m)[22m [2m{[22m
 [31mconst[39m [30mtags[39m [2m=[22m [2m[[22m[2m][22m
 [31mfor[39m [2m([22m[31mlet[39m [30mi[39m [2m=[22m [36m0[39m[2m;[22m [30mi[39m [2m<[22m [30mnames[39m[2m.[22m[36mlength[39m[2m;[22m [30mi[39m[2m+[22m[2m+[22m[2m)[22m [2m{[22m
   [30mtags[39m[2m.[22m[36mpush[39m[2m([22m[32m'[39m[32m@[39m[32m'[39m [2m+[22m [30mnames[39m[2m[[22m[30mi[39m[2m][22m[2m)[22m
 [2m}[22m
 [31mawait[39m [30mping[39m[2m([22m[30mtags[39m[2m)[22m
[2m}[22m
[31mclass[39m [36mSuperArray[39m [31mextends[39m [36mArray[39m [2m{[22m
 [31mstatic[39m [30mcore[39m [2m=[22m [36mObject[39m[2m.[22m[36mcreate[39m[2m([22m[36mnull[39m[2m)[22m
 [30mconstructor[39m[2m([22m[2m.[22m[2m.[22m[2m.[22m[30margs[39m[2m)[22m [2m{[22m [31msuper[39m[2m([22m[2m.[22m[2m.[22m[2m.[22m[30margs[39m[2m)[22m[2m;[22m [2m}[22m
 [30mbump[39m[2m([22m[30mvalue[39m[2m)[22m [2m{[22m
   [31mreturn[39m [31mthis[39m[2m.[22m[36mmap[39m[2m([22m
     [30mx[39m [2m=[22m[2m>[22m [30mx[39m [2m=[22m[2m=[22m [31mundefined[39m [2m?[22m [30mx[39m [2m+[22m [36m1[39m [2m:[22m [36m0[39m
   [2m)[22m[2m.[22m[30mconcat[39m[2m([22m[30mvalue[39m[2m)[22m
 [2m}[22m
[2m}[22m
"
`;

exports[`should work regexjs 1`] = `
"[31mexport[39m [31mconst[39m [30mtest[39m [2m=[22m [2m([22m[30mstr[39m[2m)[22m [2m=[22m[2m>[22m [32m/^/[39m[2m[[22m[36m0[39m[2m-[22m[36m5[39m[2m][22m[32m/$/g[39m[2m.[22m[30mtest[39m[2m([22m[30mstr[39m[2m)[22m
[2m// This is a super lightweight javascript syntax highlighter npm package[22m
[2m[22m[2m// This is a inline comment / <- a slash[22m
[2m[22m[2m/// <reference path="...&quot; /> // reference comment[22m
[2m[22m[2m/* This is another comment */[22m [30malert[39m[2m([22m[32m'[39m[32mgood[39m[32m'[39m[2m)[22m [2m// <- alerts[22m
[2m[22m[2m// Invalid calculation: regex and numbers[22m
[2m[22m[31mconst[39m [36m_in[39m [2m=[22m [36m123[39m [2m-[22m [32m/555/[39m [2m+[22m [36m444[39m[2m;[22m
[31mconst[39m [36m_iu[39m [2m=[22m [2m/* evaluate */[22m [2m([22m[36m19[39m[2m)[22m [2m/[22m [36m234[39m [2m+[22m [36m56[39m [2m/[22m [36m7[39m[2m;[22m
"
`;

exports[`should work jsx 1`] = `
"[31mconst[39m [30melement[39m [2m=[22m [2m([22m
  [2m<[22m[2m>[22m
    [2m<[22m[35mFood[39m
      [36mseason[39m[2m=[22m[2m{[22m[2m{[22m
        [30msault[39m[2m:[22m [2m<[22m[35mp[39m [36ma[39m[2m=[22m[2m{[22m[2m[[22m[2m{[22m[2m}[22m[2m][22m[2m}[22m [2m/>[22m
      [2m}[22m[2m}[22m[2m>[22m
    [2m</[22m[35mFood[39m[2m>[22m
    [2m{[22m[2m/* jsx comment */[22m[2m}[22m
    [2m<[22m[35mh1[39m [36mclassName[39m[2m=[22m[32m"[39m[32mtitle[39m[32m"[39m [36mdata-[39m[36mtitle[39m[2m=[22m[32m"[39m[32mtrue[39m[32m"[39m[2m>[22m[32m[39m
[32m[39m[32m      Read[39m[2m{[22m[32m'[39m[32m [39m[32m'[39m[2m}[22m[32m[39m
[32m[39m[32m      [39m[2m<[22m[35mLink[39m [36mhref[39m[2m=[22m[32m"[39m[32m/posts/first-post[39m[32m"[39m[2m>[22m[32m[39m
[32m[39m[32m        [39m[2m<[22m[35ma[39m[2m>[22m[32mthis page! - [39m[2m{[22m[36mDate[39m[2m.[22m[36mnow[39m[2m([22m[2m)[22m[2m}[22m[2m</[22m[35ma[39m[2m>[22m
      [2m</[22m[35mLink[39m[2m>[22m
    [2m</[22m[35mh1[39m[2m>[22m
  [2m</[22m[2m>[22m
[2m)[22m
"
`;

exports[`should work jsx with line numbers 1`] = `
"[2m1 [22m [31mconst[39m [30melement[39m [2m=[22m [2m([22m
[2m2 [22m   [2m<[22m[2m>[22m
[2m3 [22m     [2m<[22m[35mFood[39m
[2m4 [22m       [36mseason[39m[2m=[22m[2m{[22m[2m{[22m
[2m5 [22m         [30msault[39m[2m:[22m [2m<[22m[35mp[39m [36ma[39m[2m=[22m[2m{[22m[2m[[22m[2m{[22m[2m}[22m[2m][22m[2m}[22m [2m/>[22m
[2m6 [22m       [2m}[22m[2m}[22m[2m>[22m
[2m7 [22m     [2m</[22m[35mFood[39m[2m>[22m
[2m8 [22m     [2m{[22m[2m/* jsx comment */[22m[2m}[22m
[2m9 [22m     [2m<[22m[35mh1[39m [36mclassName[39m[2m=[22m[32m"[39m[32mtitle[39m[32m"[39m [36mdata-[39m[36mtitle[39m[2m=[22m[32m"[39m[32mtrue[39m[32m"[39m[2m>[22m[32m[39m
[2m10[22m [32m[39m[32m      Read[39m[2m{[22m[32m'[39m[32m [39m[32m'[39m[2m}[22m[32m[39m
[2m11[22m [32m[39m[32m      [39m[2m<[22m[35mLink[39m [36mhref[39m[2m=[22m[32m"[39m[32m/posts/first-post[39m[32m"[39m[2m>[22m[32m[39m
[2m12[22m [32m[39m[32m        [39m[2m<[22m[35ma[39m[2m>[22m[32mthis page! - [39m[2m{[22m[36mDate[39m[2m.[22m[36mnow[39m[2m([22m[2m)[22m[2m}[22m[2m</[22m[35ma[39m[2m>[22m
[2m13[22m       [2m</[22m[35mLink[39m[2m>[22m
[2m14[22m     [2m</[22m[35mh1[39m[2m>[22m
[2m15[22m   [2m</[22m[2m>[22m
[2m16[22m [2m)[22m
"
`;
